File: clients/rust/turk-router/tests/common/harness/raydium_amm_v4.rs
use solana_pubkey::Pubkey;
use turk_router::venues::raydium_amm_v4::{
    resolve as venue_resolve, RaydiumAmmV4Accounts, PROGRAM_ID,
};
use turk_router::VenueWindow;

use crate::common::fixture::Fixture;

pub fn resolve(fixture: &Fixture) -> VenueWindow {
    assert_eq!(
        fixture.program_id, PROGRAM_ID,
        "{}: program id",
        fixture.pool_b58
    );
    venue_resolve(RaydiumAmmV4Accounts {
        pool: fixture.slot(1).pubkey,
        base_vault: fixture.slot(3).pubkey,
        quote_vault: fixture.slot(4).pubkey,
        user_source: fixture.slot(5).pubkey,
        user_destination: fixture.slot(6).pubkey,
        payer: fixture.slot(7).pubkey,
    })
}

pub fn reachable_account_counts() -> Vec<u8> {
    let key = |byte: u8| Pubkey::new_from_array([byte; 32]);
    vec![venue_resolve(RaydiumAmmV4Accounts {
        pool: key(1),
        base_vault: key(2),
        quote_vault: key(3),
        user_source: key(4),
        user_destination: key(5),
        payer: key(6),
    })
    .account_count()]
}

File: clients/ts/tests/common/harness/raydiumAmmV4.ts
import { venues, type VenueWindow } from "../../../src/index.js";
import { slotAt, type Fixture } from "../fixture.js";
import { key } from "../keys.js";
import { assertProgram } from "./index.js";

const { PROGRAM_ADDRESS, resolve: build } = venues.raydiumAmmV4;

export function resolve(fixture: Fixture): VenueWindow {
  assertProgram(fixture, PROGRAM_ADDRESS);
  return build({
    pool: slotAt(fixture, 1).pubkey,
    baseVault: slotAt(fixture, 3).pubkey,
    quoteVault: slotAt(fixture, 4).pubkey,
    userSource: slotAt(fixture, 5).pubkey,
    userDestination: slotAt(fixture, 6).pubkey,
    payer: slotAt(fixture, 7).pubkey,
  });
}

export function reachableAccountCounts(): Array<number> {
  return [
    build({
      pool: key(1),
      baseVault: key(2),
      quoteVault: key(3),
      userSource: key(4),
      userDestination: key(5),
      payer: key(6),
    }).accountCount,
  ];
}
